tests/test_healtcheck.py

```diff
@@ -0,0 +1,169 @@
+import json
+import logging
+import random
+from time import sleep
+
+import plumbum.commands.processes
+import pytest
+from plumbum import local
+from plumbum.cmd import docker, docker_compose
+
+PROXY_TARGET_PAIRS = [
+    ("proxy_preresolve", "target"),
+    ("proxy_smtp", "target_smtp"),
+    ("proxy_without_preresolve", "target"),
+]
+
+logger = logging.getLogger()
+
+
+def _healthcheck(*args, **kwargs):
+    args = ("-f", "tests/healthcheck.yaml") + args
+    return docker_compose(*args, **kwargs)
+
+
+def _get_container_id(service_name):
+    return _healthcheck("ps", "-q", service_name).strip()
+
+
+def _get_container_id_and_ip(service_name):
+    container_id = _get_container_id(service_name)
+    container_info = json.loads(docker("inspect", container_id))
+    return (
+        container_id,
+        container_info[0]["NetworkSettings"]["Networks"][
+            "%s_simulated_outside" % local.env["COMPOSE_PROJECT_NAME"]
+        ]["IPAddress"],
+    )
+
+
+def _new_ip(target):
+    # we get the container id of the currently running target to be able to force changing ips by scaling up
+    # and then stopping the old container
+    old_container_id, old_ip = _get_container_id_and_ip(target)
+
+    # start a second instance of the target
+    _healthcheck("up", "-d", "--scale", "%s=2" % target, target)
+
+    # stop and remove the old container
+    docker("stop", old_container_id)
+    docker("rm", old_container_id)
+
+    # verify that we got a new ip (should not be able to reuse the old one)
+    new_container_id, new_ip = _get_container_id_and_ip(target)
+    assert old_container_id != new_container_id
+    assert old_ip != new_ip
+
+
+def _wait_for(proxy, message, callback, *args):
+    try:
+        while message not in callback(*args):
+            # try again in one second (to not hammer the CPU)
+            sleep(1)
+    except Exception:
+        # add additional infos to any error to make tracing down the error easier
+        logger.error("failed waiting for '%s'" % message)
+        logger.error(_healthcheck("logs", "autoheal"))
+        logger.error(_healthcheck("ps"))
+        logger.error(_healthcheck("exec", "-T", proxy, "healthcheck", retcode=None))
+        raise
+
+
+@pytest.fixture(scope="function")
+def custom_compose_project_name():
+    compose_project_name = "docker_whitelist_tests_%d" % random.randint(0, 65536)
```

Comment:
This might sound a bit nitpicking, but I'm worried that this random stuff will produce random failures if parallel tests recieve randomly the same compose project name.

I think it's better if you use the `tmp_path` fixture that pytest provides, which uses sequential numbers. Just by executing docker-compose inside of that dir should make the test more predictable. Pseudocode:

```python
def test_something(tmp_path):
    with local.cwd(tmp_path):
        docker_compose("up", "-d")  # Automatically uses non-conflicting compose project name from folder
```

Reply:
we still would need the compose_project_name or another variable (for autoheal to be limited to the current test) but yes that sounds safer

Reply:
tmp_path doesn't work with parallel tests for docker compose as default:
* /tmp/pytest-of-bigbear3001/pytest-4/popen-gw2/test_healthcheck_failing_proxy0
* /tmp/pytest-of-bigbear3001/pytest-4/popen-gw6/test_healthcheck_failing_proxy0
both of those use the same docker compose project name

Reply:
Yikes I forgot that!

I see your new implementation is more robust than both others, thanks!